testing/framework/test-framework.rst

```diff
@@ -479,67 +510,92 @@ where the path separator is a backslash).
+``print``, ``stdout`` contains the extra print output and the
+result doesn't match).
+
+Even more irritatingly, added text can cause other tests to fail and
+obscure the error you're looking for.  Say you have three different
+tests in a script excercising different code paths for the same feature,
+and the third one is unexpectedly failing. You add some debug prints to
+the affected part of scons, and now the first test of the three starts
+failing, aborting the test run before it even gets to the third test -
+the one you were trying to debug.
 
 Still, there are some techniques to help debugging.
 
 The first step should be to run the tests so the harness
 emits more information, without forcing more information into
 the test stdout/stderr which will confuse result evaluation.
-``runtest.py`` has several verbose levels which can be used
+``runtest.py`` has several levels of verbosity which can be used
 for this purpose::
 
    $ python runtest.py --verbose=2 test/foo.py
 
 You can also use the internal
 ``SCons.Debug.Trace()`` function, which prints output to
 ``/dev/tty`` on Linux/UNIX systems and ``con`` on Windows systems,
-so you can see what's going on.
+so you can see what's going on, but do not contribute to the
+captured stdout/stderr and mess up the test expectations.
 
 If you do need to add informational messages in scons code
 to debug a problem, you can use logging and send the messages
 to a file instead, so they don't interrupt the test expectations.
+Or write directly to a trace file of your choosing.
 
 Part of the technique discussed in the section
 `How to Convert Old Tests to Use Fixures`_ can also be helpful
 for debugging purposes.  If you have a failing test, try::
 
    $ PRESERVE=1 python runtest.py test/failing-test.py
 
-You can now go to the save directory reported from this run
-and invoke the test manually to see what it is doing, without
-the presence of the test infrastructure which would otherwise
-consume output you may be interested in. In this case,
+You can now go to the save directory reported from this run and invoke
+scons manually (with appropriate arguments matching what the test did)
+to see the results without the presence of the test infrastructure which
+would otherwise consume output you may be interested in. In this case,
 adding debug prints may be more useful.
 
+There are related variables ``PRESERVE_PASS``, ``PRESERVE_FAIL`` and
+``PRESERVE_NORESULT`` that preserve the directory only if the test result
+was the indicated one, which is helpful if you're trying to work with
+multiple tests showing an unusual result.
+
+From a Windows ``cmd`` shell, you will have to set the envronment
+variable first, it doesn't work on a single line like the example above for
+POSIX-style shells.
+
 
 Test infrastructure
 ===================
 
-The main test API is defined in the ``TestSCons`` class.  ``TestSCons``
+The main e2e test API is defined in the ``TestSCons`` class.  ``TestSCons``
```

Comment:
e2e -> system ?

Reply:
I don't mind using "system test" as the term, but this doc, and I think many elsewhere, have tended to use end to end, and I'm being nice and going with the existing convention? ...

Reply:
sure "end to end" vs e2e works..